### PR TITLE
Properly handle explicit "T extends Object" bounds in stub files

### PR DESCRIPTION
**Pull request by wmdietl**
This fixes typetools issue 3030.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java [hunk 1]
```diff
@@ -1816,6 +1815,18 @@ private void annotateTypeParameters(
...
+                        && param.getTypeBound().get(0).getAnnotations().isEmpty()
+                        && paramType
+                                .getUpperBound()
+                                .getUnderlyingType()
+                                .toString()
+                                .contentEquals("java.lang.Object")) {
+                    // If there is an explicit "T extends Object" type parameter bound,
+                    // treat it like an explicit use of "Object" in code.
+                    AnnotatedTypeMirror ub = atypeFactory.getAnnotatedType(Object.class);
+                    paramType.getUpperBound().addAnnotations(ub.getAnnotations());
```
Comment (cpovirk) [inline framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:1828]: Could we need to first remove any existing annotations? Granted, it's far less likely that such annotations would exist, particularly in the JDK :) But in principle, stubs _could_ be used to override a source `<T extends @Nullable Object>` with plain `<T extends Object>`.

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java [hunk 2]
```diff
@@ -1816,6 +1815,18 @@ private void annotateTypeParameters(
...
+                if (param.getTypeBound().size() == 1
+                        && param.getTypeBound().get(0).getAnnotations().isEmpty()
+                        && paramType
+                                .getUpperBound()
+                                .getUnderlyingType()
+                                .toString()
+                                .contentEquals("java.lang.Object")) {
+                    // If there is an explicit "T extends Object" type parameter bound,
+                    // treat it like an explicit use of "Object" in code.
+                    AnnotatedTypeMirror ub = atypeFactory.getAnnotatedType(Object.class);
```
Comment (cpovirk) [inline framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:1827]: In theory, it seems like we'd want to apply any defaults associated with this location, like those set by `@DefaultQualifier`. In practice, that probably doesn't come up. And it sounds potentially hard to do :)

**Review by cpovirk — APPROVED** [review]
Thanks, this is even better :)

I left a couple comments about ways in which this PR might theoretically not handle every edge case. But I'm not concerned about those in practice. I bring them up mostly in case they hint at some other problem that actually _could_ matter in practice.

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java (hunk 1 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java]: In https://github.com/eisop/checker-framework/pull/247/commits/c8239840b7f28fdcf50cc934f098d86e56d1ac99 I've added a test where the bytecode contains an explicit Nullable bound and the same four combinations of stubfiles. The tests work correctly, but I still changed the `addAnnotations` to `replaceAnnotations`, for consistency with the rest of the file.

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java (hunk 2 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:1827]: I was also a bit worried about that, but at the moment I don't see anything that would suggest a `@DefaultQualifier` in an astub file (and therefore in the annotated JDK) has any effect.
That would be another unfortunate mismatch between `.java` files and `.astub` files, but is probably hard to fix. Let's punt on that until we actually run into that issue.

Comment (cpovirk) [reply framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:1828]: (We've actually used `@DefaultQualifier` in a stub file successfully but only with the goal of affecting the defaults that the checker applies to the actual source code. I don't know that we've ever tried to use it to affect what's written in the stub file, which is what we're talking about here.)

File: framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java (hunk 1 above)
Comment (cpovirk) [reply framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java]: Thanks for checking! I can believe it works; I didn't try to dig deeply enough to see where the original objects came from and whether they might have already been annotated.

**Pull request merged**